soc: st: stm32: flash prefetch on U3, spread CONFIG_STM32_FLASH_PREFETCH, flash cache on U0

etienne-lms opened the pull request:
Enable flash prefetch support on STM32U3x that was missing.

Use `CONFIG_STM32_FLASH_PREFETCH` for enabling flash prefetch on all applicable STM32 SoCs, for consistency. Add hidden option `CONFIG_HAS_STM32_FLASH_PREFETCH` for this purpose.

Enable flash instruction cache on STM32U0x that was missing.

etienne-lms committed:
soc: st: stm32u3: add flash prefetch

Implement flash prefetch for STM32U3xx SoCs.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

erwango commented on soc/st/stm32/stm32c0x/Kconfig.defconfig:
```diff
@@ -7,6 +7,9 @@ if SOC_SERIES_STM32C0X
 
 rsource "Kconfig.defconfig.stm32c0*"
 
+config STM32_FLASH_PREFETCH
+	default y
+
```
Since this is not series specific, why not moving this in `soc/st/stm32/Kconfig.defconfig` ?

mathieuchopstm replied:
Having an `HAS_FLASH_PREFETCH` symbol would be cleaner, since the option would not be visible on SoCs where it is not available. (It could be made conditional based on flash controller `compatible` to avoid explicitly listing supported series too, but that'd be another PR)

etienne-lms commented:
By the way, I'd like to add a bit of description for existing `STM32_FLASH_PREFETCH` that has no `help` section.
https://github.com/zephyrproject-rtos/zephyr/blob/c80ace50f4ae6e8ef2bc59933c356604811dd910/soc/st/stm32/Kconfig#L50-L52

mathieuchopstm commented on soc/st/stm32/Kconfig:
```diff
@@ -55,8 +55,19 @@ config SWJ_ANALOG_PRIORITY
+	help
+	  Enable flash prefetch. For power efficiency, tt is recommended to
+	  be enabled when flash is accessed through a cache memory. Flash
+	  prefetch also improves a bit performances.
```
I think:
```suggestion
	  prefetch also improves performance a bit.
```

kbidani commented:
> Comments addressed.

thnx ^^

etienne-lms committed:
soc: st: stm32: use CONFIG_STM32_FLASH_PREFETCH where applicable

Use CONFIG_STM32_FLASH_PREFETCH config option to enable flash prefetch
in C0/F0/F1/F2/F3/F4/F7/G4/L0/L1/L4/U0/U3/WB/WBA/WBL for consistency
with other SoCs that use this configuration switch, default enabled
at SoC default config level.

Add SoC hidden config option HAS_STM32_FLASH_PREFETCH enabled for
SoCs that support the feature. STM32_FLASH_PREFETCH is default
enabled for all SoC that have HAS_STM32_FLASH_PREFETCH unless target
specific constraints (as for G0Bx/G0Cx SoCs).

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
soc: st: stm32u0: enable flash instruction cache

Enable flash instruction cache on U0 SoCs.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

gautierg-st reviewed (approved):
Has an ST internal ticket been opened to have `LL_FLASH_EnablePrefetch` available in System LL for U3 instead of calling the HAL function?
